Add first solution3-2

Nigma-Ks opened the pull request.

Nigma-Ks committed:
Add first solution3-2

Lozov-Petr commented on hw3_2/hw3-2/Program.cs:
```diff
@@ -0,0 +1,452 @@
+
+                using (StreamReader reader = new StreamReader(path))
+                {
+                    if ((text = await reader.ReadToEndAsync()) == null)
```
```suggestion
                    if ((text = reader.ReadToEnd()) == null)
```
Зачем синхронно вызывать асинхронную операцию?)

Lozov-Petr commented on hw3_2/hw3-2/Program.cs:
```diff
@@ -0,0 +1,452 @@
+                }
+
+                bool isEmpty;
+                List<Byte[]> listOfCodes = lzw.LzwCompression(text, out isEmpty);
```
```suggestion
                (List<Byte[]> listOfCodes, bool isEmpty) = lzw.LzwCompression(text);
```
Не используйте `out` параметры. Вместо этого просто верните пару

Lozov-Petr commented on hw3_2/hw3-2/Program.cs:
```diff
@@ -0,0 +1,452 @@
+                    }
+                }
+
+                Console.WriteLine($"Compression ratio: {4 * listOfCodes.Count / (float)(2 * text.Length)}");
```
```suggestion
                Console.WriteLine($"Compression ratio: {listOfCodes.Count / text.Length}");
```
Не понял назначение этих констант. Кажется, что достаточно размер результата компрессии поделить на размер исходного файла. Если я не прав, напишите тут.
И да, если используете магические константы, смысл которых неочевиден - выносите их в именованные константы.

Lozov-Petr replied:
А, это количества бит для инта и для символа.
Точно надо в константы.
А ещё лучше, вынести вычисление коэффициента в отдельный метод.

Lozov-Petr commented on hw3_2/hw3-2/Program.cs:
```diff
@@ -0,0 +1,452 @@
+
+                bool isEmpty;
+                bool isCorrect;
+                string decompressedText = lzw.LzwDecompression(GroupBytesByFour(allBytes), out isEmpty, out isCorrect);
```
Опять же, верните тройку результатов, вместо использования `out` аргументов.

Lozov-Petr commented on hw3_2/hw3-2/Program.cs:
```diff
@@ -0,0 +1,452 @@
+
+                StreamWriter unzippedFileWriter = new StreamWriter(newFileName);
+                unzippedFileWriter.Write(decompressedText);
+                unzippedFileWriter.Close();
```
```suggestion
                using var unzippedFileWriter = new StreamWriter(newFileName);
                unzippedFileWriter.Write(decompressedText);
```
1) Не пишите тип, когда он очевиден из право части.
2) using тут очень к месту.

Lozov-Petr commented on hw3_2/hw3-2/Program.cs:
```diff
@@ -0,0 +1,452 @@
+
+        public static List<Byte[]> GroupBytesByFour(Byte[] arrayOfBytes)
+        {
+            List<Byte[]> listOfBytesByFour = new List<byte[]>();
```
```suggestion
           var listOfBytesByFour = new List<byte[]>();

```

Lozov-Petr commented on hw3_2/hw3-2/Program.cs:
```diff
@@ -0,0 +1,452 @@
+
+        private void AddAllSymbolsToStorage()
+        {
+            for (int i = 0; i < 256; i++)
```
256 вынесите в константу

Lozov-Petr commented on hw3_2/hw3-2/Program.cs:
```diff
@@ -0,0 +1,452 @@
+        public int Size {
+            get => _size;
+            private set { _size = value; }
+        }
```
```suggestion
        public int Size {
            get;
            private set;
        }
```
Вы написали абсолютно стандартную реализацию для `get` и `set`. Для такого случая есть более краткая форма.
Такая форма также неявно создаст приватное поле, где будет храниться значение.

Lozov-Petr commented on hw3_2/hw3-2/Program.cs:
```diff
@@ -0,0 +1,452 @@
+
+            isEmptyText = false;
+            AddAllSymbolsToStorage();
+            int textForCompressionLenght = textForCompression.Length;
```
```suggestion
            int textForCompressionLength = textForCompression.Length;

```

Lozov-Petr commented on hw3_2/hw3-2/Program.cs:
```diff
@@ -0,0 +1,452 @@
+            {
+                string symbol = "";
+                symbol += (char)i;
+                _storage.Add(symbol);
```
```suggestion
                _storage.Add(new string((char)i, 1))
```
Так покороче будет

Lozov-Petr commented on hw3_2/hw3-2/Program.cs:
```diff
@@ -0,0 +1,452 @@
+            {
+                string symbol = "";
+                symbol += (char)i;
+                list.Add(symbol);
```
```suggestion
                _storage.Add(new string((char)i, 1))
```

Lozov-Petr commented on hw3_2/hw3-2.Tests/UnitTest1.cs:
```diff
@@ -0,0 +1,85 @@
+        if (resultLenght != 11 || isEmpty) //Lenght of arrayOfRightCodes
+        {
+            Assert.IsTrue(false);
+        }
```
```suggestion
        Assert.IsTrue(resultLenght != 11 || isEmpty);
```

Lozov-Petr commented on hw3_2/hw3-2.Tests/UnitTest1.cs:
```diff
@@ -0,0 +1,85 @@
+            if (BitConverter.ToUInt32(result[i]) != arrayOfRightIntCodes[i])
+            {
+                Assert.IsTrue(false);
+            }
```
```suggestion
            Assert.IsTrue(BitConverter.ToUInt32(result[i]) != arrayOfRightIntCodes[i]);
```

Lozov-Petr commented on hw3_2/hw3-2.Tests/UnitTest1.cs:
```diff
@@ -0,0 +1,85 @@
+namespace hw3_2.Tests;
+
```
```suggestion

using NUnit.Framework;

```
А у вас без `using` тесты запускались?

Lozov-Petr commented on hw3_2/hw3-2/Program.cs:
```diff
@@ -0,0 +1,452 @@
+                    Console.WriteLine("Impossible to decompress. Wrong codes");
+                }
+
+                string newFileName = path.Replace(".zipped", "");
```
`Replace` заменит все вхождения ".zipped" в строке.
Например, если мы захотим сжать файл дважды, то получим имя "name.zipped.zipped". А при первом расжатии потеряем оба расширения.

Lozov-Petr commented on hw3_2/hw3-2/Program.cs:
```diff
@@ -0,0 +1,452 @@
+            isNewString = InsertString(_root, element); //not null because constructs with no null root
+            if (isNewString)
+            {
+                Size = ++Size;
```
```suggestion
                ++Size;
```
Сам по себе инкремент уже увеличивает значение в переменной на единицу.

Lozov-Petr replied:
В данном решении используются коды фиксированной длины. И я не нашел в реализации проверки, что коды ещё не кончились. Если этого нет - добавьте. Если есть, напишите, где.

Lozov-Petr reviewed (commented):
3

Lozov-Petr commented:
В данном решении используются коды фиксированной длины. И я не нашел в реализации проверки, что коды ещё не кончились. Если этого нет - добавьте. Если есть, напишите, где.

Nigma-Ks committed:
1. разнесла классы по файлам
2. изменила внутреннее устройство хранилища в соответсвии с изменениями во втором домашнем задании

Nigma-Ks committed:
1. Добавила функцию подсчета сжатия
2. Добавила функцию для составления имени нового файла при разжатии
3. Добавила проверку на переполнение хранилища

Nigma-Ks committed:
Прошлый коммит не прошел

Nigma-Ks committed:
fa-> false
